lib/suse/connect/client.rb

```diff
@@ -37,12 +37,11 @@ def register!
       def deregister!
         raise SystemNotRegisteredError unless registered?
         if @config.product
-          raise BaseProductDeactivationError if @config.product == Zypper.base_product
-          service = deactivate_product @config.product
-          remove_or_refresh_service(service)
-          Zypper.remove_release_package @config.product.identifier
-          print_success_message @config.product, action: 'Deregistered'
+          deregister_product(@config.product)
         else
+          Zypper.installed_products.each do |product|
+            deregister_product(product) unless product == Zypper.base_product
+          end
```

Comment:
I think we probably want to do this *after* `@api.deregister(system_auth)`, because if the API call fails for whatever reason (network error, etc) then the system will be in a strange state.

Reply:
Ah, well we're also calling the API in `deregister_product`, so I guess it's ok actually.

Reply:
I'm wondering if we need integration tests for this. I would guess yes?
I'm not sure what our "policy" is regarding integration tests.

Reply:
the latest changes (amended to previous commit) add an integration test